capsules/src/proximity.rs

```diff
@@ -0,0 +1,310 @@
+//!
+//!let proximity = static_init!(
+//!   capsules::proximity::ProximitySensor<'static>,
+//!   capsules::proximity::ProximitySensor::new(apds9960 , board_kernel.create_grant(&grant_cap)));
+//!
+//!kernel::hil::sensors::ProximityDriver::set_client(apds9960, proximity);
+//! ```
+
+use kernel::hil;
+use kernel::ReturnCode;
+use kernel::{AppId, Callback, Driver, Grant};
+
+/// Syscall driver number.
+use crate::driver;
+pub const DRIVER_NUM: usize = driver::NUM::Proximity as usize;
+
+#[derive(Default)]
+pub struct App {
+    callback: Option<Callback>,
+    subscribed: bool,
+    enqueued_command_type: ProximityCommand,
+    lower_proximity: u8,
+    upper_proximity: u8,
+}
+
+#[derive(Clone, Copy, PartialEq)]
+pub enum ProximityCommand {
+    Exists = 0,
+    ReadProximity = 1,
+    ReadProximityOnInterrupt = 2,
+}
+
+impl Default for ProximityCommand {
+    fn default() -> Self {
+        ProximityCommand::Exists
+    }
+}
+
+#[derive(Default)]
+pub struct Thresholds {
+    lower: u8,
+    upper: u8,
+}
+
+pub struct ProximitySensor<'a> {
+    driver: &'a dyn hil::sensors::ProximityDriver<'a>,
+    apps: Grant<App>,
+}
+
+impl<'a> ProximitySensor<'a> {
+    pub fn new(
+        driver: &'a dyn hil::sensors::ProximityDriver<'a>,
+        grant: Grant<App>,
+    ) -> ProximitySensor<'a> {
+        ProximitySensor {
+            driver: driver,
+            apps: grant,
+        }
+    }
+
+    fn enqueue_command(
+        &self,
+        command: ProximityCommand,
+        arg1: usize,
+        arg2: usize,
+        appid: AppId,
+    ) -> ReturnCode {
+        // Enqueue command by saving command type, args, appid within app struct in grant region
+        let r: ReturnCode = self
+            .apps
+            .enter(appid, |app, _| {
+                // Return busy if same app attempts to enqueue second command before first one is callbacked
+                if app.subscribed {
+                    return ReturnCode::EBUSY;
+                }
+
+                if command == ProximityCommand::ReadProximityOnInterrupt {
+                    app.lower_proximity = arg1 as u8;
+                    app.upper_proximity = arg2 as u8;
+                }
+
+                app.subscribed = true; // enqueue
+                app.enqueued_command_type = command;
+
+                ReturnCode::SUCCESS
+            })
+            .unwrap_or_else(|err| err.into());
+
+        if r == ReturnCode::EBUSY {
+            return ReturnCode::EBUSY;
+        }
+
+        // Only run command if it is only one in queue otherwise we wait for callback() for last run command to trigger another command to run
+        let mut num_commands: u8 = 0;
+
+        for cntr in self.apps.iter() {
+            cntr.enter(|app, _| {
+                if app.subscribed {
+                    num_commands += 1;
+                }
+            });
+        }
+        if num_commands == 1 {
+            self.run_next_command();
+        }
```

Comment:
If app1 issues a `ReadProximityOnInterrupt` command, then app2 issues a `ReadProximity` command, doesn't this force the "instantaneous" reading to not occur until after the reading that is waiting on an interrupt occurs?

Reply:
This is a great observation.  I have decided to put a "ReturnCode::EBUSY" if the currently running command is an interrupt command and an app is issuing a "ReadProximity" command.  I have not found a good solution general to all chips where the interrupt command can be briefly interrupted by a "instantaneous" read command and then resume afterwards.  If you think this is a bad fix then I would be happy to consider your input here.

Reply:
Hmm, that is a little unfortunate because it means that any app configuring a `ReadProximityWithInterrupt` command effectively takes an indefinite lock on the sensor, which violates the Tock threat model. What do you think of the following approach:

Currently, you acknowledge that multiple `ReadProximityWithInterrupt` commands can be processed simultaneously, and that multiple `ReadProximity` commands can be processed simultaneously. 

If a `ReadProximity` command is issued while a `ReadProximityWithInterrupt` command is "running", the capsule should  change the running command to `ReadProximity` and tell the driver to execute an immediate read instead. The outstanding command(s) will still be enqueued (the apps marked as subscribed), and thus are not forgotten. Also, the callback logic should be changed:

- when a callback arrives for a `ReadProximity` command, still check it against all apps with outstanding `ReadProximityWithInterrupt` commands, and if the proximity value falls within the interrupt-causing range deliver that callback to that client and clear its outstanding command. This helps ensure that proximity readings which arrive during the period when interrupts are disabled still can be delivered to clients that were waiting for interrupts.

Whenever a `ReadProximity` command arrives, it should only be possible for `ReadProximityWithInterrupt` commands to be outstanding, (as any `ReadProximity` callback will remove unsubscribe all `ReadProximity` clients). Also, the driver always handles all `ReadProximityWithInterrupt` commands simultaneously via the call to `find_thresholds()`, so the interrupt detection should only be briefly disabled while the sensor processes the immediate read and services the interrupt. This approach does allow for the small possibility of missed interrupts if the sensor detects the desired proximity value in the time period after the immediate reading is taken and before the capsule has re-enabled interrupts. However this should be quite unlikely, and I think is probably worth making the application interface non-locking.

Reply:
This is a pretty good resolution. Ill try the implementation and some testing.

Reply:
The callback logic is much cleaner now.  Anytime a reading comes in (no matter for which command it is in response to), the callback logic will check against all outstanding apps and if the app is waiting on an interrupt then it will check against the thresholds.  Any app waiting for a immediate reading or `ReadProximity` should be able to get the command regardless.  Thus, there was really no need for the `command_type` argument in the callback.

I have also implemented the condition in the `enqueue_command` function which checks if the command running is an interrupt command and then substitutes it with the `ReadProximity` command.

Reply:
awesome!